### PR TITLE
Some bugfixes, documentation, and cleanup

### PR DESCRIPTION
**Pull request by arbennett**
This pull request has 8 main components:
1. It fixes docker runs.
2. It fixes bugs output control for adding options with spaces in the names and the statistic ordering
3. It fixes a bug where running multiple ensembles lead to errors.
4. It makes dependency versioning less strict, which allows newer versions of libraries
5. It adds some documentation to internal functions.
6. Moves from attribute access to item access for options (`s.local_param_info.albedoMax` => `s.local_param_info['albedoMax']`
7. Adds the ability to reset simulations by backing up the initial configuration
8. Minimizes data transfer during the launching of Ensembles by adding an additional hidden `initialize` step.

### CODE REVIEW

File: pysumma/simulation.py [hunk 1]
```diff
@@ -64,22 +93,25 @@ def gen_summa_cmd(self, processes=1, prerun_cmds=[],
...
+                    prerun_cmds=None, startGRU=None, countGRU=None, iHRU=None,
+                    freq_restart=None, progress=None):
+        """Start a docker simulation"""
+        run_cmd = self._gen_summa_cmd(run_suffix, processes, prerun_cmds,
+                                      startGRU, countGRU, iHRU,
+                                      freq_restart, progress)
+        run_cmd = run_cmd.replace(self.executable, '/code/bin/summa.exe')
 
         fman_dir = os.path.dirname(self.manager_path)
         settings_path = self.manager.settings_path.value
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: self.manager doesn't have settings_path attribute.
So change to 
`settings_path = self.manager.get_value("settings_path")`
`input_path = self.manager.get_value("input_path")`
`output_path = self.manager.get_value("output_path")`

File: pysumma/simulation.py [hunk 2]
```diff
@@ -88,132 +120,105 @@ def run_docker(self, run_suffix=None, processes=1,
         cmd = ''.join(['docker run -v {}:{}'.format(fman_dir, fman_dir),
                        ' -v {}:{}'.format(settings_path, settings_path),
                        ' -v {}:{}'.format(input_path, input_path),
-                       ' -v {}:{}'.format(output_path, output_path),
-                       '/bin/bash -c "',
+                       ' -v {}:{} '.format(output_path, output_path),
+                       " --entrypoint '/bin/bash' ",
+                       self.executable,
```
Comment (DavidChoi76) [inline pysumma/simulation.py:125]: need to define `self.executable` like `self.executable = "uwhydro/summa:sopron_2018"`

File: pysumma/simulation.py [hunk 3]
```diff
@@ -64,22 +93,25 @@ def gen_summa_cmd(self, processes=1, prerun_cmds=[],
...
-        self._status = 'Running'
+        self.status = 'Running'
 
-    def run_docker(self, run_suffix=None, processes=1,
-                   prerun_cmds=[], startGRU=None, countGRU=None, iHRU=None,
-                   freq_restart=None, progress=None):
-        run_cmd = self.gen_summa_cmd(processes, prerun_cmds,
-                                     startGRU, countGRU, iHRU,
-                                     freq_restart, progress)
+    def _run_docker(self, run_suffix, processes=1,
```
Comment (DavidChoi76) [inline pysumma/simulation.py:107]: image

This is a docker running error.

File: pysumma/simulation.py (hunk 2 above)
Comment (arbennett) [reply pysumma/simulation.py:125]: You should be able to give `"uwhydro/summa:sopron_2018"` as the `executable` when constructing the `Simulation` instance. So like this:

```
s = ps.Simulation('uwhydro/summa:sopron_2018', file_manager)
s.run('docker', 'docker_test')
```

File: pysumma/simulation.py (hunk 3 above)
Comment (arbennett) [reply pysumma/simulation.py:107]: Could you provide more information about how you set up the simulation?

File: pysumma/simulation.py (hunk 1 above)
Comment (arbennett) [reply pysumma/simulation.py]: This should now be fixed.

**Pull request merged**